onnxruntime/core/graph/runtime_optimization_record.h

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <vector>
+#include <string>
+
+#include "core/graph/basic_types.h"
+
+/* Runtime optimization limitations
+ *
+ * A runtime optimization cannot optimize a node which was created by another runtime optimization.
+ *
+ * This constraint is believed to be acceptable now because:
+ * - Runtime optimizations are expected to be limited to replaying certain extended-level optimizations that fuse ONNX
+ *   nodes into contrib op nodes.
+ *   - Thus, they should not be handling any nodes that are produced by other runtime optimizations.
+ * - Layout optimizations are expected to run directly in a minimal build instead of being replayed with runtime
+ *   optimizations.
+ *
+ * If this constraint needs to be lifted at some point, note these considerations:
+ * 1. Nodes produced by runtime optimizations need to be tracked more carefully when loading the runtime optimizations.
+ *    The current approach is to identify nodes by their indices. A node's index in the graph depends on when the node
+ *    gets added to the graph. The ordering of node additions when saving runtime optimizations may be different from
+ *    the ordering when loading them. There needs to be a way to refer to the same runtime optimization-produced node
+ *    at save and load time.
+ * 2. There is an intermediate state between a runtime optimization and a later runtime optimization that depends on a
+ *    node produced by the first. The ORT format model should contain the initial (not intermediate) graph state plus
+ *    the runtime optimizations. There needs to be a way to make the result of a runtime optimization visible to later
+ *    runtime optimizations and also save the initial graph state.
+ */
+
+namespace onnxruntime {
+
+/** Struct to serialize the node indices in an ORT format model.
+Use NodesToOptimize::EmptyNodeIndex for nullptr entries in the vectors for missing optional inputs
```

Comment:
NodesToOptimize::EmptyNodeIndex is std::numeric_limits<NodeIndex>::max() which is the max of size_t, which is platform dependent

Reply:
Should we use std::numeric_limits<uint32_t>::max() as we serialize NodeIndex to the flatbuffer using uint32?

Reply:
changed the value to uint32_t max. we'll have to be careful about not storing larger NodeIndex values in these structures